### DISCUSSION TITLE
[8주차] 링크 계층 6.1~3

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Dijkstra-Employ/Network-Study#16] [Q&A]
8주차 링크 계층 6.1~3에 해당하는 질의응답입니다.

**Comment by Starlight258** [comment]
# Q1 
`링크 계층`이 제공하는 서비스는 무엇인가요?

# Q2
`링크 계층`은 `하드웨어`와 `소프트웨어` 모두에서 구현되어있습니다. 
`링크 계층`의 기능을 `구현된 위치`에 따라 분류하여 설명해주세요.

**Reply by H-sooyeon** [reply]
> 링크 계층이 제공하는 서비스는 무엇인가요?

링크 계층의 기본 서비스는 단일 통신 링크상으로 데이터그램을 한 노드에서 인접 노드로 이동시키는 것이지만, 서비스의 세부 사항은 링크에 적용되는 특정 링크 계층 프로토콜에 따른다.
- 프레임화
  - 데이터그램을 링크상으로 전송하기 전에 링크 계층 프레임에 캡슐화한다.
- 링크 접속
  - MAC 프로토콜은 링크상으로 프레임을 전송하는 규칙을 명시한다.
- 신뢰적 전달
  - 오류율이 높은 링크에서 주로 사용되며, 링크 계층 프로토콜은 오류가 발생한 링크에서 정정한다.
- 오류 검출과 정정
  - 오류 검출은 하드웨어로 구현되며 오류 정정은 프레임 안의 오류 검출뿐만 아니라 프레임의 어느 곳에서 오류가 발생했는지 찾아낼 수 있다는 점을 제외하면 오류 검출과 비슷하다.

<br />

> 링크 계층은 하드웨어와 소프트웨어 모두에서 구현되어있습니다. 링크 계층의 기능을 구현된 위치에 따라 분류하여 설명해주세요.

대부분의 링크 계층은 네트워크 인터페이스 컨트롤러(NIC)로 알려진 네트워크 어댑터에 구현된다.
네트워크 어댑터의 중심에는 링크 계층 컨트롤러가 있으며, 링크 계층 컨트롤러는 링크 계층 서비스(프레임화, 링크 접속, 오류 검출 등)의 대다수가 구현되어 있는 단일의 특수 용도 칩이다.
링크 계층 컨트롤러의 기능 대부분은 하드웨어로 구현되며, 일부 기능은 호스트 CPU에서 실행되는 소프트웨어로 구현되어 있다.
즉, 링크 계층은 하드웨어와 소프트웨어의 조합이다.

**Reply by Starlight258** [reply]
> # Q1
> `링크 계층`이 제공하는 서비스는 무엇인가요?
> 
- 프레임화: 링크 계층 프로토콜은 `네트워크 계층 데이터그램`을 링크상으로 전송하기 전에, `링크 계층 프레임`에 캡슐화한다.
- 링크 접속: 하나의 `브로드캐스트 링크`를 여러 노드가 공유하는 경우(다중 접속 문제), MAC 프로토콜은 **여러 노드로부터의 프레임 전송을 조정**한다.
- 신뢰적 전달: **네트워크 계층 데이터그램은 링크상에서 오류 없이 전달**된다. 
-  오류 검출과 정정: 송신 노드에서 프레임에 `오류 검출 비트`를 설정하고, 수신 노드에서 `오류 검출`을 수행함으로써 가능하다.

> # Q2
> `링크 계층`은 `하드웨어`와 `소프트웨어` 모두에서 구현되어있습니다. `링크 계층`의 기능을 `구현된 위치`에 따라 분류하여 설명해주세요.
- 하드웨어: 링크 계층  제어기의 기능 대부분은 하드웨어로 구현된다.
  - 링크 계층은 `네트워크 인터페이스 카드 (NIC)` 라고 알려진 `네트워크 어댑터`에 구현된다.
  - 네트워크 어댑터의 중심에는 `링크 계층 제어기`가 있다.
- 소프트웨어: 일부 링크 계층 기능은 `호스트 CPU에서 실행되는 소프트웨어`에 구현되어 있다.
  -  제어기가 기본적인 작업을 수행하지만, 소프트웨어가 더 복잡하고 유연한 처리를 담당한다.

**Comment by H-sooyeon** [comment]
## Q1

브로드캐스트 링크에서의 문제와 이를 보완하기 위해 나온 프로토콜에 대해 설명해주세요.

## Q2

링크 계층에서의 오류 검출 방법에 대해 간단하게 설명해주세요.

**Reply by Starlight258** [reply]
> ## Q1
> 브로드캐스트 링크에서의 문제와 이를 보완하기 위해 나온 프로토콜에 대해 설명해주세요.
> 
- 문제: 2개 이상의 노드가 동시에 프레임을 전송하는 경우, 충돌이 일어나 낭비된다.
- 해결 : `다중 접속 프로토콜`: 브로드캐스트 채널로 보내는 노드들의 전송을 조정하기 위한 프로토콜
  - `채널 분할 프로토콜`, `랜덤 접속 프로토콜`, `순번 프로토콜` 로 분류된다.
 
> ## Q2
> 링크 계층에서의 오류 검출 방법에 대해 간단하게 설명해주세요.
- 데이터 오류를 검출하는 기술로 `패리티 검사`, `체크섬`, `순환중복검사`를 수행한다.
- `패리티 검사`: 수신자는 1의 개수를 계산하여 오류 발생 여부를 알 수 있다. 가장 단순한 형태이다.
- `체크섬`: 수신자는 수신 데이터 합의 `1의 보수`를 취한 뒤, 그 결과가 모두 `1`인 비트로 구성되어있는지 검사한다.
  - 한 비트라도 `0`이면 오류가 발생한 것이다.
  - 오버헤드가 적다.
- `순환중복검사` : 주어진 데이터 D에 대해 r개의 추가 비트 R을 뒤에 덧붙인다. `d+r 비트`가 `모듈러-2 연산`을 수행하여 `G`로 나누었을때, 나머지가 존재하면 오류가 발생한 것으로 간주한다.

**Comment by choiyoung69** [comment]
## Q1 
오늘날 네트워크에서 많이 사용되는 오류 검출 기술에는 순환 중복 검사가 있습니다. 이 검사의 작동 방식에 대해서 간단하게 설명해주세요.


## Q2
다중 접속 프로토콜에는 세 가지 종류가 있습니다. 각각 프로토콜의 특징에 대해 설명해주세요

**Reply by H-sooyeon** [reply]
> 오늘날 네트워크에서 많이 사용되는 오류 검출 기술에는 순환 중복 검사가 있습니다. 이 검사의 작동 방식에 대해서 간단하게 설명해주세요.

CRC는 이진 다항식을 기반으로 작동한다.
데이터를 전송할 때, 이 데이터를 다항식으로 해석하여 나눗셈 연산을 수행한다.

- 송신자가 데이터를 생성 다항식으로 나눠 나머지(CRC 코드)를 계산하여 데이터와 함께 전송한다.
- 수신자가 데이터를 받아 같은 방식으로 나눗셈을 수행하여 나머지를 확인한다.
- 나머지가 0이면 오류 없음, 나머지가 0이 아니면 오류가 발생한 것으로 판단한다.

<br />

> 다중 접속 프로토콜에는 세 가지 종류가 있습니다. 각각 프로토콜의 특징에 대해 설명해주세요
1. 채널 분할 프로토콜
채널을 공유하는 모든 노드가 브로드캐스트 채널의 대역폭을 분할할 수 있게 해주는 두 가지 기술인 시분할 다중화(TDM)와 주파수 분할 다중화(FDM)으로 나뉜다.

2. 랜덤 접속 프로토콜
랜덤 접속 프로토콜에서 전송 노드는 항상 채널의 최대 전송률인 Rbps로 전송하며 충돌이 생기면 충돌과 관련된 각 노드는 프레임이 충돌 없이 전송될 때까지 자신의 프레임을 계속해서 재전송한다.
또한, 프레임을 재전송하기 전에 충돌과 관련된 각 노드는 자신만의 랜덤 지연 시간을 선택하며 그 시간동안 기다린다.
알로하 프로토콜, CSMA 프로토콜이 있다.

3. 순번 프로토콜
랜덤 접속 프로토콜은 단 하나의 노드만이 활성이면 그 노드가 Rbps의 처리율을 가진다는 특성을 가지지만, 또 다른 요구 특성인 M개의 노드가 활성이면 각 노드가 거의 R/Mbps의 처리율을 갖는다는 특성을 가지지는 않는다.
이를 보완하기 위해 순번 프로토콜이 개발되었으며, 폴링 프로토콜, 토큰 전달 프로토콜이 있다.

**Reply by Starlight258** [reply]
> ## Q1
> 오늘날 네트워크에서 많이 사용되는 오류 검출 기술에는 순환 중복 검사가 있습니다. 이 검사의 작동 방식에 대해서 간단하게 설명해주세요.
> 
- 주어진 데이터 D에 대해 r개의 추가 비트 R을 뒤에 덧붙인다.
-  `d+r 비트`가 `모듈러-2 연산`을 수행하여 `G`로 나누었을때, 나머지가 존재하면 오류가 발생한 것으로 간주한다.
> 참고) `모듈러-2 연산` : 덧셈의 올림이나 뺄셈의 빌림이 없음, 피연산자를 비트별로 `xor` 한 것과 같다.

> ## Q2
> 다중 접속 프로토콜에는 세 가지 종류가 있습니다. 각각 프로토콜의 특징에 대해 설명해주세요
- 채널 분할 프로토콜: 채널을 공유하는 모든 노드가 브로드캐스트 채널의 대역폭을 분할하게 해주는 기술
  - 1. 시분할 다중화(`TDM`)
  - 2. 주파수 분할 다중화(`FDM`)
 - 랜덤 접속 프로토콜
    - 전송 노드는 항상 채널의 최대 전송률인 `R bps`로 전송한다.
   - 충돌이 생기면 충돌 발생 노드는 프레임이 충돌 없이 전송될 때까지 자신의 프레임을 계속해서 `재전송`한다.  
   - 충돌이 생긴 노드는 자신만의 랜덤 지연 시간을 선택한다.
   - 슬롯 알로하, 알로하가 있다.

- 순번 프로토콜
  - 폴링 프로토콜: 노드 중 하나를 `마스터 노드`로 지정한다.
	- 마스터 노드는 각 노드를 `라운드 로빈` 방식으로 폴링한다. 
  - 토큰 전달 프로토콜: 노드들 사이에 정해진 순서대로 `토큰`이 전달된다.
	- 노드가 토큰을 수신하면, 전송할 프레임이 있을 때만 토큰을 붙잡는다.
	- 효율이 매우 높다.

**Comment by Hanjaemo** [comment]
### Q1
전송 계층에서 종단 간 신뢰적 데이터 전달을 보장하는데, 왜 링크 계층에서도 신뢰적 데이터 전달 서비스를 제공하는 것일까요?
그리고 둘은 어떤 차이가 있을까요?

### Q2
CSMA/CD와 CSMA/CA의 차이점은 무엇이며, 각각의 장단점을 설명해 주세요.

**Reply by H-sooyeon** [reply]
> 전송 계층에서 종단 간 신뢰적 데이터 전달을 보장하는데, 왜 링크 계층에서도 신뢰적 데이터 전달 서비스를 제공하는 것일까요? 그리고 둘은 어떤 차이가 있을까요?

전송 계층은 네트워크의 종단 간 데이터를 정확하게 전달하는 데 중점을 둔다.
여기서 신뢰적 전달은 네트워크 전체를 대상으로 하며, 데이터가 여러 중간 노드를 거쳐서 이동할 때 발생하는 오류나 손실을 수신 측에서 종합적으로 처리한다.

링크 계층은 인접한 노드 간의 데이터 전송을 신뢰적으로 보장한다.
이는 두 장치가 물리적으로 연결된 단이 네트워크 구간(이더넷 연결)에서 오류 없이 데이터를 송수신하는 것을 목표로 한다.
링크 계층은 매체의 특성상 발생할 수 있는 노이즈, 신호 손실, 충돌 등을 관리한다.

링크 계층이 구간 내의 오류를 신속하게 처리함으로써 전송 계층에서 신뢰성을 보장하는 부담을 줄일 수 있다.
즉, 링크 계층이 로컬에서 발생하는 오류를 미리 처리해주기 때문에 전송 계층이 전체 경로에서 발생하는 복잡한 문제에 집중할 수 있다.

<br />

> CSMA/CD와 CSMA/CA의 차이점은 무엇이며, 각각의 장단점을 설명해 주세요.

`CSMA/CD`는 장치가 네트워크 매체를 통해 데이터를 전송하기 전에  링크를 사용중인지 확인하고, 충돌이 발생한 후에 이를 감지하여 충돌 시 전송을 중단한 뒤 랜덤한 시간 후에 재전송한다.
CSMA/CD는 유선 네트워크(이더넷)에서 주로 사용된다.

[ 장점 ]
- 충돌이 발생한 후 빠르게 감지하여 처리하므로 효율적인 충돌 처리 매커니즘이다.
- 간단하고 구현이 귑다.

[ 단점 ]
- 충돌 발생 후에 처리하므로, 충돌로 인한 대역폭 낭비가 발생한다.
- 충돌이 많아질 경우 네트워크 성능이 저하된다.

`CSMA/CA`는 충돌을 사전에 회피하는 방식이다.
데이터를 전송하기 전에 노드가 데이터를 전송 중인지 확인하고, 전송하려는 의도를 미리 알림으로써 충돌을 피하려고 한다.
CSMA/CA는 무선 네트워크(Wi-Fi)에서 주로 사용된다.
무선 환경에서는 충돌을 감지하기 어려우므로, 충돌을 사전에 회피하는 방식이 더 효과적이다.

[ 장점 ]
- 충돌을 사전에 피하려 하므로 충돌로 인한 대역폭 낭비가 줄어든다.
- 여러 장치가 동시에 통신할 때도 비교적 효율적이다.
- 무선 환경에서 안정적으로 작동하며, 특히 전파 범위 내에서 발생할 수 있는 충돌을 줄이는 데 유리하다.

[ 단점 ]
- 충돌을 피하기 위해 RTS/CTS 신호와 같은 추가적인 제어 메시지를 사용하므로 전체적으로 오버헤드가 증가한다.
- 네트워크의 전송 지연이 발생할 수 있다.
- 네트워크 사용량이 적을 때는 오히려 CSMA/CD보다 효율이 떨어질 수 있다.

**Reply by Starlight258** [reply]
> ### Q1
> 전송 계층에서 종단 간 신뢰적 데이터 전달을 보장하는데, 왜 링크 계층에서도 신뢰적 데이터 전달 서비스를 제공하는 것일까요? 그리고 둘은 어떤 차이가 있을까요?
> 
- 전송 계층이나 애플리케이션 계층에서 데이터를 `종단간에 재전송`하는 것과 달리, 링크 계층의 신뢰적 전달은 `오류가 발생한 링크에서 오류를 정정`한다.
- 오류율이 높은 링크에서 주로 사용된다.
- 낮은 비트율을 가진 링크에서는 **불필요한 오버헤드**가 될 수도 있다.
- 대다수 유선 링크 계층 프로토콜은 신뢰적 전달 서비스를 제공하지 않는다.

> ### Q2
> CSMA/CD와 CSMA/CA의 차이점은 무엇이며, 각각의 장단점을 설명해 주세요.
- CSMA/CD는 주로 유선 이더넷에서 사용되고, CSMA/CA는 무선 네트워크(예: Wi-Fi)에서 사용된다.
- `CSMA/CD`
  - 장점:  충돌 감지시 즉시 전송을 중단하므로 성능을 향상시킬 수 있으며 구현이 단순하다.
  - 단점: 무선 환경에서 충돌 감지가 어렵다. 
    - 무선 환경에서는 신호의 강도가 거리에 따라 급격히 감소하므로  다른 노드의 약한 신호를 감지하기 어렵다.
- `CSMA/CA`
  - 장점: 충돌을 감지하는 대신 충돌을 피하는 데 중점을 두어 충돌 가능성을 사전에 줄인다.
    - 무선 환경에 적합하다.
  - 단점: 추가적인 제어 프레임(RTS/CTS)으로 인한 오버헤드가 발생한다.
    - 구현이 복잡하다.
